pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -498,6 +503,43 @@ func (a *Actuator) ensureAnnotation(ctx context.Context, machine *machinev1beta1
 	return true, a.client.Update(ctx, machine)
 }
 
+// ensureProviderID adds the providerID to the Machine spec, if it does
+// not already exist.  Also sets the corresponding ID on the related
+// Node when it becomes referenced via the Status.NodeRef
+func (a *Actuator) ensureProviderID(ctx context.Context, machine *machinev1beta1.Machine, host *bmh.BareMetalHost) error {
+	providerID := "baremetalhost:///" + host.Namespace + "/" + host.Name
+	existingProviderID := machine.Spec.ProviderID
+	if existingProviderID == nil || *existingProviderID != providerID {
+		log.Printf("Setting ProviderID %s for machine %s.", providerID, machine.Name)
+		machine.Spec.ProviderID = &providerID
+		err := a.client.Update(ctx, machine)
+		if err != nil {
+			log.Printf("Failed to update machine ProviderID, error: %s", err.Error())
+			return err
+		}
+	}
+
+	node, err := a.getNodeByMachine(ctx, machine)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Printf("Not setting Node ProviderID - Node does not yet exist for Machine %s", machine.Name)
+			return nil
```

Comment:
What causes the controller to try again if the node isn't there? Does the controller watch Nodes as well as Machines?

Reply:
It does watch Nodes, [looks for the machine annotation](https://github.com/openshift/cluster-api-provider-baremetal/blob/59fae9d/pkg/manager/wrapper/node_mapper.go) on the Node, and queues the corresponding Machine for reconciliation. The annotation itself gets [added by the nodelink controller](https://github.com/openshift/machine-api-operator/blob/2a985cf91608b290374fbf5db366ce0577714d2b/pkg/controller/nodelink/nodelink_controller.go#L233).

Reply:
Thanks, TIL